### PR TITLE
Use Pydantic to verify System schemas

### PR DESCRIPTION
**Pull request by amaslenn**
## Summary
Use Pydantic to verify System schemas.

1. Removed System Parsers, added `parse_system()` into `Parser` class.
2. Refactored `SlurmSystem`: introduced `SlurmPartition` and `SlurmGroup` classes to manage node lists.
3. All Systems are supported now, including K8s.
4. Added a new mode for verifying system schemas: `verify-systems`:
    ```sh
    cloudai --mode verify-systems --tests-dir s --test-templates-dir d --system-config conf/common/system
    ```

## Test Plan
CI

## Additional Notes
— 

## A message for release notes
We are working on schema improvements to simplify configs management and make them verifiable. This will help ensure that configs are correct before expensive runs on real hardware. Today we are enabling it for System configs.
1. Added new command for verifying the configs: `cloudai --mode verify-systems`. `--system-config` can be a file or a directory to verify all configs in the directory.
2. Slurm system config format was updated to take advantage of TOML features:
    ```toml
    [partitions]
    [partitions.partition_1]
    name = "partition_1"
    nodes = ["node-[001-100]"]
    
    [partitions.partition_2]
    name = "partition_2"
    nodes = ["node-[101-200]"]
    ```
    is now
    ```toml
    [[partitions]]
    name = "partition_1"
    nodes = ["node-[001-100]"]
    
    [[partitions]]
    name = "partition_2"
    nodes = ["node-[101-200]"]
    ```
    The same is for groups inside partitions.
3. System parser objects were removed, this functionality is now handled by Pydantic.

### CODE REVIEW

File: README.md
Comment (TaekyungHeo) [inline README.md:122]: Can we specify directory names based on our current directory structure to maintain consistency with other sections of this README file?

File: conf/common/system/example_slurm_cluster.toml
Comment (TaekyungHeo) [inline conf/common/system/example_slurm_cluster.toml:28]: I agree with this change because it is a more TOML-native way to describe sections. However, we need to communicate with users to ensure they are aware of this change. @srinivas212, please review.

File: src/cloudai/__main__.py
Comment (TaekyungHeo) [inline src/cloudai/__main__.py:277]: Do we print proper error messages upon failure? Based on my experience with CloudAI users, it is always better to provide more descriptive information, warning, and error messages for debugging purposes.

File: conf/common/system/example_slurm_cluster.toml
Comment (TaekyungHeo) [reply conf/common/system/example_slurm_cluster.toml:28]: @amaslenn, could you please create a PR for CloudAIX as well? The system configuration files will need to be updated accordingly once the PR is merged.

File: src/cloudai/__main__.py
Comment (amaslenn) [reply src/cloudai/__main__.py:277]: Yes, here is how it looks like:
```bash
cloudai --mode verify-systems --tests-dir s --test-templates-dir d --system-config conf/common/system/example_slurm_cluster.toml
[INFO] Verifying conf/common/system/example_slurm_cluster.toml...
[ERROR] Validation error: {'type': 'missing', 'loc': ('name',), 'msg': 'Field required', 'input': {'scheduler': 'slurm', 'install_path': './install', 'output_path': './results', 'default_partition': 'partition_1', 'mpi': 'pmix', 'gpus_per_node': 8, 'ntasks_per_node': 8, 'partitions': [{'name': 'partition_1', 'nodes': ['node-[001-100]'], 'groups': [{'name': 'group_1', 'nodes': ['node-[001-025]']}, {'name': 'group_2', 'nodes': ['node-[026-050]']}, {'name': 'group_3', 'nodes': ['node-[051-075]']}, {'name': 'group_4', 'nodes': ['node-[076-100]']}]}, {'name': 'partition_2', 'nodes': ['node-[101-200]']}], 'global_env_vars': {'NCCL_IB_GID_INDEX': '3', 'NCCL_IB_TIMEOUT': '20', 'NCCL_IB_QPS_PER_CONNECTION': '4', 'MELLANOX_VISIBLE_DEVICES': '0,3,4,5,6,9,10,11', 'CUDA_VISIBLE_DEVICES': '0,1,2,3,4,5,6,7'}}}
[ERROR] Field 'name': Field required
```

File: README.md
Comment (amaslenn) [reply README.md]: Sure, updated.

**Review by srivatsankrishnan — COMMENTED** [review]
LGTM overall. The test plan only relies on CI/Unit tests. Is it also possible to test this on a simple slurm system in addition to CI?  This is a big change in my opinion. 

Without making modification to the system schemas in private repo, we cannot ideally use this feature. Should wait before this PR is also ready: https://github.com/Mellanox/cloudaix/pull/37?

**Comment by srivatsankrishnan** [comment]
`What kind of tests would you like to add? Currently, we load our slurm example system config only.`

We can try a simple verification (gpt or grok or even nccl single node test) that Daria is running in one additional internal cluster? Just an additional test other than CI.  I assume we can modify the system schema she is using to use the Pydantic. I think your draft PR [here](https://github.com/Mellanox/cloudaix/pull/37/files#diff-00e0f17d1fa519d3fad3ceb4035003998b1ae6ad7508d24e682bda7db3ded467) can still be used to test this PR?

`This PR requires changes in Cloud AI, it cannot be ready until we make a release.`

Right, we usually install the a local copy of cloudAI to test features we add. We don't have to wait for a release right? Also, I assume we are also planning to add new features and want to ensure it doesn't break any features we have in the pipeline. Most of the changes in this PR looks transparent and I assume it won't break anything.  If you have one example validation on simple verification other than CI that will helpful.

**Comment by amaslenn** [comment]
> We can try a simple verification (gpt or grok or even nccl single node test) that Daria is running in one additional internal cluster? Just an additional test other than CI. ... 

This all comes to the single question: do we trust our tests or not.
I tend to trust our set of tests, it is pretty comprehensive.

My question to you @srivatsankrishnan is what do you think is missing in our CI so we don't rely on real HW verification as it is slow and costly. Our unit tests could provide much better coverage, we just need to add more tests where needed.

What additional coverage will real HW run provide? And is there a reason not to cover it with unit tests?

**Review by TaekyungHeo — APPROVED** [review]

**Comment by srivatsankrishnan** [comment]
You are making a big change to the system schema and how it's getting processed within cloudAI and it's not about trusting the CI test or not. This PR needs to be aligned with your other draft PR and it would be good if we can test this those changes. You already have changes that one need to make in system schema. I think even a simple sleep test on a real slurm system would be helpful. If we have to rebase of this version, I want to have a validated system schema to begin with.

**Comment by amaslenn** [comment]
> I think even a simple sleep test on a real slurm system would be helpful. 

I still don't get how this is better compared to all our CI tests. But anyway, I run one test:
```bash
cloudai --mode run --system-config conf/common/system/jazz.toml --test-templates-dir conf/common/test_template --tests-dir conf/common/test --test-scenario conf/common/test_scenario/sleep.toml
[INFO] System configuration file: conf/common/system/jazz.toml
[INFO] Test templates directory: conf/common/test_template
[INFO] Tests directory: conf/common/test
[INFO] Test scenario file: conf/common/test_scenario/sleep.toml
[INFO] Output directory: None
[WARNING] No JsonGenStrategy found for TestTemplateParser and SlurmSystem
[WARNING] No JsonGenStrategy found for TestTemplateParser and SlurmSystem
[WARNING] No JsonGenStrategy found for TestTemplateParser and SlurmSystem
[WARNING] No JsonGenStrategy found for TestTemplateParser and SlurmSystem
[WARNING] No JsonGenStrategy found for TestTemplateParser and SlurmSystem
[INFO] System Name: hpchead
[INFO] Scheduler: slurm
[INFO] Test Scenario Name: test_scenario_example
[INFO] Checking if test templates are installed.
[INFO] Test Scenario: test_scenario_example

Section Name: Tests.1
  Test Name: sleep_10
  Description: sleep_10
  No dependencies
[INFO] Initializing Runner
[INFO] Creating SlurmRunner
[INFO] Starting test scenario execution.
[INFO] Starting test: Tests.1
[INFO] Running test: Tests.1
[INFO] Executing command for test Tests.1: sbatch results/test_scenario_example_2024-09-11_09-57-39/Tests.1/0/cloudai_sbatch_script.sh
[INFO] Job completed: Tests.1
[INFO] All test scenario results stored at: results/test_scenario_example_2024-09-11_09-57-39
[WARNING] Skipping directory 'results/test_scenario_example_2024-09-11_09-57-39/Tests.1/0' for test 'sleep_10'
[INFO] All test scenario execution attempts are complete. Please review the 'debug.log' file to confirm successful completion or to identify any issues.
```

(for the rest of your message, please see my comment in Cloud AI X PR)

**Pull request merged**
